File: Tarea 04/procesamientodigital.py
import sounddevice as sd
import numpy as np
import matplotlib.pyplot as plt
import psycopg2
import base64
import wave

# Conexión a la base de datos
conn = psycopg2.connect("dbname=postgres user=postgres password=koala host=localhost port=5432")
cur = conn.cursor()

opcion = 0
while opcion != 5:
    print("\n========== Menú Principal ==========")
    print("1) Grabar audio")
    print("2) Reproducir audio")
    print("3) Graficar señal")
    print("4) Graficar espectro de frecuencia")
    print("5) Salir")

    try:
        opcion = int(input("Ingrese su elección: "))
    except ValueError:
        opcion = 0

    if opcion == 1:
        duracion = int(input("Ingrese la duración de la grabación (segundos): "))
        print("Comenzando la grabación...")
        audio = sd.rec(int(duracion * 44100), samplerate=44100, channels=1, dtype='int16')
        sd.wait()
        print("Grabación finalizada.")

        nombre_archivo = "audio.wav"
        with wave.open(nombre_archivo, 'wb') as wf:
            wf.setnchannels(1)
            wf.setsampwidth(2)
            wf.setframerate(44100)
            wf.writeframes(audio.tobytes())

        print("Archivo guardado correctamente.")

        with open(nombre_archivo, "rb") as f:
            audio_bytes = f.read()
        audio_base64 = base64.b64encode(audio_bytes).decode('utf-8')
        cur.execute("INSERT INTO audios (nombre, audio) VALUES (%s, decode(%s, 'base64'))",
                    (nombre_archivo, audio_base64))
        conn.commit()
        print("Audio guardado en la base de datos.")

    elif opcion == 2:
        try:
            print("Reproduciendo audio...")
            with wave.open("audio.wav", 'rb') as wf:
                data = wf.readframes(wf.getnframes())
                audio = np.frombuffer(data, dtype='int16')
                sd.play(audio, wf.getframerate())
                sd.wait()
        except:
            print("Error al reproducir el audio.")

    elif opcion == 3:
        try:
            with wave.open("audio.wav", 'rb') as wf:
                data = wf.readframes(wf.getnframes())
                audio = np.frombuffer(data, dtype='int16')
                tiempo = np.linspace(0, len(audio) / wf.getframerate(), num=len(audio))
            plt.plot(tiempo, audio)
            plt.xlabel("Tiempo (s)")
            plt.ylabel("Amplitud")
            plt.title("Audio")
            plt.show()
        except:
            print("Error al graficar la señal.")

    elif opcion == 4:
        try:
            with wave.open("audio.wav", 'rb') as wf:
                data = wf.readframes(wf.getnframes())
                audio = np.frombuffer(data, dtype='int16')
                fs = wf.getframerate()
            N = len(audio)
            freqs = np.fft.rfftfreq(N, 1/fs)
            espectro = np.abs(np.fft.rfft(audio))**2
            plt.plot(freqs, 10 * np.log10(espectro))
            plt.xlabel("Frecuencia (Hz)")
            plt.ylabel("Potencia (dB)")
            plt.title("Espectro de Frecuencia")
            plt.show()
        except:
            print("Error al graficar el espectro.")

    elif opcion == 5:
        print("Saliendo del programa...")
    else:
        print("Opción no válida.")

File: Tarea 04/procesamientodigital.m
#Comprobar si es Octave
if (exist('OCTAVE_VERSION','builtin')~=0)
  pkg load signal;
  pkg load database;
end

#Conexión a la base de datos
conn = pq_connect(setdbopts('dbname','postgres','host','localhost','port','5432','user','postgres','password','koala'));

#Variable para el menu principal
opcion = 0;

while opcion ~= 5
  disp('========== Menú Principal ==========');
  disp('1) Grabar audio');
  disp('2) Reproducir audio');
  disp('3) Graficar señal');
  disp('4) Graficar espectro de frecuencia');
  disp('5) Salir');

  opcion = input('Ingrese su eleccion: ');

  switch opcion
      case 1
        #Grabar audio
          try
              duracion = input('Ingrese la duracion de la grabacion en segundos: ');
              disp('Comenzando la grabacion...');
              recObj = audiorecorder;
              recordblocking(recObj, duracion);
              disp('Grabación finalizada.');
              data = getaudiodata(recObj);
              nombreArchivo = 'audio.wav';
              audiowrite(nombreArchivo, data, recObj.SampleRate);
              disp('Archivo de audio guardado correctamente');

              #Guardar en la base de datos
              fid = fopen(nombreArchivo, 'rb');
              audioBytes = fread(fid, '*uint8');
              fclose(fid);
              query = sprintf("INSERT INTO audios (nombre, audio) VALUES ('%s', decode('%s','base64'))", ...
                  nombreArchivo, base64_encode(audioBytes));
              pq_exec_params(conn, query);
              disp('Audio guardado en la base de datos.');

          catch
              disp(['Error al grabar el audio: ', e.message]);
          end

      case 2
        #Reproducir audio
          try
              disp('Reproduciendo audio...');
              [data, fs] = audioread('audio.wav');
              sound(data, fs);
          catch
              disp('Error al reproducir el audio.');
          end

      case 3
      #Graficar señal
          try
              disp('Graficando señal...');
              [data, fs] = audioread('audio.wav');
              tiempo = linspace(0, length(data)/fs, length(data));
              plot(tiempo, data);
              xlabel('Tiempo (s)');
              ylabel('Amplitud');
              title('Audio');
          catch
              disp('Error al graficar el audio.');
          end

      case 4
      #Grafica espectro de frecuencia
          try
              disp('Graficando espectro de frecuencia...');
              [audio, fs] = audioread('audio.wav');  #Lee la señal desde el archivo .wav
              N = length(audio);  #Numero de muestras de la señal
              f = linspace(0, fs/2, N/2+1);  #Vector de frecuencia
              ventana = hann(N);  #Ventana de Hann para reducir el efecto de las discontinuidades al calcular la FFT
              Sxx = pwelch(audio, ventana, 0, N, fs);  #Densidad espectral de potencia
              plot(f, 10*log10(Sxx(1:N/2+1)));  #Grafica el espectro de frecuencia en dB
              xlabel('Frecuencia (Hz)');
              ylabel('Densidad espectral de potencia (dB/Hz)');
              title('Espectro de Frecuencia de la señal grabada');
          catch
              disp('Error al graficar el audio.');
          end

      case 5
      #Salir
          disp('Saliendo del programa...');
      otherwise
          disp('Opción no válida.');
  end
end

